storage/fanout.go

```diff
@@ -188,142 +206,146 @@ func (f *fanoutAppender) Rollback() (err error) {
-		queriers:       filtered,
+		merge:          (&seriesMergerAdapter{VerticalSeriesMergeFunc: merge}).Merge,
+		primary:        newGenericQuerierFrom(primary),
+		secondaries:    filtered,
 		failedQueriers: make(map[genericQuerier]struct{}),
 		setQuerierMap:  make(map[genericSeriesSet]genericQuerier),
 	}}
 }
 
-// NewMergeChunkQuerier returns a new ChunkQuerier that merges results of chkQuerierSeries chunk queriers.
-// NewMergeChunkQuerier will return NoopChunkQuerier if no chunk queriers are passed to it,
-// and will filter NoopQuerieNoopChunkQuerierrs from its arguments, in order to reduce overhead
-// when only one chunk querier is passed.
-func NewMergeChunkQuerier(primaryQuerier ChunkQuerier, queriers []ChunkQuerier, merger VerticalChunkSeriesMergerFunc) ChunkQuerier {
-	filtered := make([]genericQuerier, 0, len(queriers))
-	for _, querier := range queriers {
+// NewMergeChunkQuerier returns a new ChunkQuerier that merges results of given primary and slice of secondary chunk queriers.
+// The difference between primary and rest is as follows:
+// * If the primary querier returns an error, the query fails.
+// * For secondaries it just returns warnings.
+// In case of overlaps between the data given by all secondaries, merge will be used.
+// TODO(bwplotka): Currently merge will compact overlapping chunks with bigger chunk, without limit. Split it: https://github.com/prometheus/tsdb/issues/670
+func NewMergeChunkQuerier(primary ChunkQuerier, secondaries []ChunkQuerier, merge VerticalChunkSeriesMergeFunc) ChunkQuerier {
+	filtered := make([]genericQuerier, 0, len(secondaries))
+	for _, querier := range secondaries {
 		if _, ok := querier.(noopChunkQuerier); !ok && querier != nil {
 			filtered = append(filtered, newGenericQuerierFromChunk(querier))
 		}
 	}
 
 	if len(filtered) == 0 {
-		return primaryQuerier
-	}
-
-	if primaryQuerier == nil && len(filtered) == 1 {
-		return &chunkQuerierAdapter{filtered[0]}
+		return primary
 	}
 
 	return &chunkQuerierAdapter{&mergeGenericQuerier{
-		mergeFunc:      (&chunkSeriesMergerAdapter{VerticalChunkSeriesMergerFunc: merger}).Merge,
-		primaryQuerier: newGenericQuerierFromChunk(primaryQuerier),
-		queriers:       filtered,
+		merge:          (&chunkSeriesMergerAdapter{VerticalChunkSeriesMergeFunc: merge}).Merge,
+		primary:        newGenericQuerierFromChunk(primary),
+		secondaries:    filtered,
 		failedQueriers: make(map[genericQuerier]struct{}),
 		setQuerierMap:  make(map[genericSeriesSet]genericQuerier),
 	}}
 }
 
 // Select returns a set of series that matches the given label matchers.
-func (q *mergeGenericQuerier) Select(sortSeries bool, hints *SelectHints, matchers ...*labels.Matcher) (genericSeriesSet, Warnings, error) {
-	if len(q.queriers) == 1 {
-		return q.queriers[0].Select(sortSeries, hints, matchers...)
-	}
-
+// NOTE: mergeGenericQuerier selects always return series sorted no matter if sorting was not required from caller.
```

Comment:
I can see people depending on this when they shouldn't as it's a function comment, I'd put this inside the function as an implementation note.

Reply:
ack